IronFrame/Utilities/FirewallManager.cs

```diff
@@ -43,6 +50,87 @@ public void OpenPort(int port, string name)
             firewallPolicy.Rules.Add(firewallRule);
         }
 
+        internal static string GetFormattedLocalUserSid(string windowsUsername)
+        {
+            var ntaccount = new NTAccount("", windowsUsername);
+            var sid = ntaccount.Translate(typeof(SecurityIdentifier)).Value;
+            return String.Format(CultureInfo.InvariantCulture, "D:(A;;CC;;;{0})", sid);
+        }
+
+        /// <summary>
+        /// Remove all firewall rules that match the given name
+        /// 
+        /// The implementation is a bit hacky. There is no way determine 
+        /// how many rules match a given name without iterating through 
+        /// the entire collection of rules. This method seems cheaper 
+        /// but uses exceptions to signal that no more rules match the given
+        /// name.
+        /// </summary>
+        /// <param name="userName"></param>
+        public void RemoveAllFirewallRules(string userName)
+        {
+            var firewallPolicy = (INetFwPolicy2)Activator.CreateInstance(Type.GetTypeFromProgID(NetFwPolicy2ProgID));
+            var rules = firewallPolicy.Rules;
+            try
+            {
+                // keep deleting until an exception is thrown
+                while (true)
+                {
+                    // rules.Item will throw an exception if the rule isn't found 
+                    rules.Item(userName);
+                    rules.Remove(userName);
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                // ignore the exception
+            }
+        }
+
+        public void CreateFirewallRule(string windowsUserName, FirewallRuleSpec firewallRuleSpec)
+        {
+            var protocol = firewallRuleSpec.Protocol;
+            if (protocol == Protocol.All)
+            {
+                _CreateFirewallRule(windowsUserName, Protocol.Udp, firewallRuleSpec);
+                _CreateFirewallRule(windowsUserName, Protocol.Tcp, firewallRuleSpec);
+            }
+            else
+            {
+                _CreateFirewallRule(windowsUserName, protocol, firewallRuleSpec);    
+            }
+        }
+
+        private void _CreateFirewallRule(string windowsUserName, Protocol proto, FirewallRuleSpec firewallRuleSpec)
+        {
+            var firewallPolicy = (INetFwPolicy2)Activator.CreateInstance(Type.GetTypeFromProgID(NetFwPolicy2ProgID));
+
+            // This type is only avaible in Windows Server 2012
+            var rule = ((INetFwRule3)Activator.CreateInstance(Type.GetTypeFromProgID(NetFwRuleProgID)));
+
+            rule.Name = windowsUserName;
```

Comment:
Is this the rule name that someone would see in the Firewall UI ?

I think we use the container ID in existing code that creates firewall policies.  It would be good to be consistent.